runtime/parachains/src/session_info/migration.rs

```diff
@@ -20,3 +20,32 @@ use frame_support::traits::StorageVersion;
 
 /// The current storage version.
 pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+pub mod v3 {
+	use crate::{session_info, shared};
+	use frame_support::{pallet_prelude::Weight, traits::OnRuntimeUpgrade};
+	use frame_system::Config;
+	use primitives::vstaging::ExecutorParams;
+	use sp_core::Get;
+
+	pub struct MigrateToV3<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config + session_info::pallet::Config> OnRuntimeUpgrade for MigrateToV3<T> {
+		fn on_runtime_upgrade() -> Weight {
+			// Bootstrap session executor params with the default ones if no parameters for the
+			// current session are in storage. `ExecutorParams::default()` is supposed to generate
+			// EXACTLY the same set of parameters the previous implementation used in a hard-coded
+			// form. This supposed to only run once, when upgrading from pre-parametrized executor
+			// code.
+			let mut weight = T::DbWeight::get().reads(2);
+			let session_index = <shared::Pallet<T>>::session_index();
+			if <session_info::Pallet<T>>::session_executor_params(session_index).is_none() {
+				session_info::pallet::SessionExecutorParams::<T>::insert(
+					&session_index,
+					ExecutorParams::default(),
+				);
+				weight += T::DbWeight::get().writes(1);
+			}
+			weight
+		}
+	}
+}
```

Comment:
@eskimor @ordian it's my first attempt to implement a migration which is bootstrapping current session's executor params, as proposed by @ordian. I'm pretty sure I've screwed it up but I had to start with something.

My main doubts are:
1. Do we need to bump `StorageVersion` in case when no actual storage format migration is done, but a new `StorageMap` is added and needs to be bootstrapped?
2. Is it okay to use `V3` as a migration version definition? It goes well along with both `SessionInfo` structure version and runtime API version but indeed doesn't relate to either of them. Should I use independent versioning here? Which one? `MigrateToV1` looks strange and `MigrateToV2` doesn't work because nothing defines what `V2` is.
3. I'm total noob in using weights so I also have some doubts if I made it the right way.

Reply:
I think what you've implemented makes sense and works.
1. You're required to bump it, but it's a good practice to do that and add a version check that ensures this runtime upgrade code is idempotent and only executed once.
2. The version here should reflect the `StorageVersion` speicific to this module, not `RuntimeApi` version or any other version. They are orthogonal. In theory one could add storage in one runtime upgrade and runtime api to query it in another.
3. Yes, that's correct.

One thing we might want to add is `pre_upgrade` and `post_upgrade` methods gated by `try-runtime` features. Example: https://github.com/paritytech/substrate/blob/master/frame/fast-unstake/src/migrations.rs